File: src/main/java/com/oycm/algorithm/ba/modify/Solution_12.scala
package com.oycm.algorithm.ba.modify

object Solution_12 {

  /**
   * 75. 颜色分类
   * https://leetcode.cn/problems/sort-colors/description/
   *
   * 整数 0、 1 和 2 分别表示红色、白色和蓝色。
   * 原地对 nums 进行排序同颜色的元素相邻，并按照红色、白色、蓝色顺序排列。
   *
   * @param nums 包含红色、白色和蓝色 的数组
   */
  def sortColors(nums: Array[Int]): Unit = {
    /*
    只有 0, 1, 2 元素的数组, 原地进行升序排序
    两次交换：把 0 全部交换到左边，把 2 全部交换到右边

    时间复杂度 O(n)
    */
    var i0 = 0
    for (i <- nums.indices) {
      if (nums(i) == 0) {
        val temp = nums(i0)
        nums(i0) = nums(i)
        nums(i) = temp
        i0 += 1
      }
    }
    val n = nums.length
    var i2 = n - 1
    var i = n - 1
    while (i >= i0) {
      if (nums(i) == 2) {
        val temp = nums(i2)
        nums(i2) = nums(i)
        nums(i) = temp
        i2 -= 1
      }
      i -= 1
    }
    //println(nums.toSeq)
  }

  def answer(nums: Array[Int]): Unit = {
    /*
    题解思路：
    如何在一个有序数组 a = [0,0,1,1,2,2], 插入一个 元素 0, 并使 数组 a 还保持有序
    暴力做法： 在数组的最左边插入 0, 其它所有元素右移一位, 时间复杂度是 O(n)
    [0,0,1,1,2,2]
    [0,0,0,1,1,2,2]
    对比插入后的表现
    1. a(2) 变成 0
    2. a(4) 变成 1
    3. 末尾新增 2, a(6) = 2

    修改过程中，如何知道修改的下标？
    1. 维护 a 中 0 的数量, 即要修改成 0 的位置, a0 = 2, 即 a(a0) = 0
    2. 维护 a 中 0 和 1 的数量, 即要修改成 1 的位置, a1 = 4, 即 a(a1) = 1
    3. 末尾新增的位置记为 i, 即 a(i) = 2

    注意，如果 a = [1], a 中没有 2, 上面的第 3 步就错了，结果变成了 a = [0,2]
    如果 a = [0], a 中没有 2 和 1, 上面的第 2, 3 步就错了，结果变成了 a = [0,2]
    按照这个逻辑修改，需要进行多种 if-else 判断
    如果倒过来修改呢？是不是就能解决这个问题。
    1. 新增位置记为 i, 即 a(i) = 2
    2. a(a1) = 1
    3. a(a0) = 0
    假设数组长 为 n, 0 的数量记为 a0, 1 的数量记为 a1, 2 的数量记为 a2
    如果 a2 > 0 正反修改都没有问题
    a2 == 0, a1 即要插入的位置 i, a(i) 先修改成 2, 再修改成 1, 完全没有问题
    如果 a1 == a2 == 0, 逻辑同上

    如果在有序数组 a 中插入的是 1
    1. a(i) = 2
    2. a(a1) = 1

    如果在有序数组 a 中插入的是 2
    1. a(i) = 2

    */
    var a0 = 0
    var a1 = 0
    for (i <- nums.indices) {
      val x = nums(i)
      // 0 | 1 | 2 的第 1 步
      nums(i) = 2
      // 0 或 1 的第 2 步
      if (x <= 1) {
        nums(a1) = 1
        a1 += 1
      }
      // 0 的第三步
      if (x == 0) {
        nums(a0) = 0
        a0 += 1
      }
    }
  }

  def main(args: Array[String]): Unit = {
    sortColors(Array(2, 0, 2, 1, 1, 0))
    answer(Array(2, 0, 2, 1, 1, 0))
  }

}

File: src/main/java/com/oycm/algorithm/i/search/Solution_12.java
package com.oycm.algorithm.i.search;

import java.util.Arrays;

public class Solution_12 {

    /**
     * 473. <a href="https://leetcode.cn/problems/matchsticks-to-square/description/">火柴拼正方形</a>
     *
     * @param matchsticks 整数数组, matchsticks[i] 是第 i 个火柴棒的长度
     * @return
     */
    public boolean makesquare(int[] matchsticks) {
        /*
        要用 所有的火柴棍 拼成一个正方形。你 不能折断 任何一根火柴棒，但你可以把它们连在一起，而且每根火柴棒必须 使用一次 。
         */
        /*
        先计算所有长度和 sum, sum % 4 != 0 肯定无法拼凑
        接下来就看是否能把 matchsticks 分成 4 组 和为 sum / 4 的子集
         */
        int sum = 0;
        for (int x : matchsticks) {
            sum += x;
        }
        if (sum % 4 != 0) {
            return false;
        }
        Arrays.sort(matchsticks);
        for (int l = 0, r = matchsticks.length - 1; l < r; l++, r--) {
            int temp = matchsticks[l];
            matchsticks[l] = matchsticks[r];
            matchsticks[r] = temp;
        }
        return dfs(0, matchsticks, new int[4], sum / 4);
    }

    public boolean dfs(int i, int[] nums, int[] path, int target) {
        if (i == nums.length) {
            // 前面已经判断过总合，能进到之类，说明已经分好了
            return true;
        }
        for (int j = 0; j < path.length; j++) {
            /*
            优化：如果当前桶的值和前一个一样，前面选如果符合要求就不会到这里，所以这里再选也是无效的
             */
            if (j > 0 && path[j] == path[j - 1]) continue;

            if (path[j] + nums[i] > target) continue;
            path[j] += nums[i];
            if (dfs(i + 1, nums, path, target)) {
                return true;
            }
            path[j] -= nums[i];
        }

        return false;
    }

}

class Solution_698 {

    /**
     * 698. <a href="https://leetcode.cn/problems/partition-to-k-equal-sum-subsets/description/">划分为k个相等的子集</a>
     *
     * @param nums
     * @param k
     * @return
     */
    public boolean canPartitionKSubsets(int[] nums, int k) {
        int sum = 0;
        for (int x : nums) {
            sum += x;
        }
        if (sum % k != 0) {
            return false;
        }
        Arrays.sort(nums);
        for (int l = 0, r = nums.length - 1; l < r; l++, r--) {
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
        }
        return dfs(0, nums, new int[k], sum / k);
    }

    public boolean dfs(int i, int[] nums, int[] path, int target) {
        if (i == nums.length) {
            // 前面已经判断过总合，能进到之类，说明已经分好了
            return true;
        }
        for (int j = 0; j < path.length; j++) {
            /*
            优化：如果当前桶的值和前一个一样，前面选如果符合要求就不会到这里，所以这里再选也是无效的
             */
            if (j > 0 && path[j] == path[j - 1]) continue;

            if (path[j] + nums[i] > target) continue;
            path[j] += nums[i];
            if (dfs(i + 1, nums, path, target)) {
                return true;
            }
            path[j] -= nums[i];
        }

        return false;
    }
}

class Solution_2305 {

    /**
     * 2305. <a href="https://leetcode.cn/problems/fair-distribution-of-cookies/description/">公平分发饼干</a> 1887
     *
     * @param cookies 整数数组 cookies ，其中 cookies[i] 表示在第 i 个零食包中的饼干数量。length [2, 8]
     * @param k       等待分发零食包的孩子数量
     * @return
     */
    public int distributeCookies(int[] cookies, int k) {
        /*
        所有 零食包都需要分发，在同一个零食包中的所有饼干都必须分发给同一个孩子，不能分开。
        分发的不公平程度 定义为单个孩子在分发过程中能够获得饼干的最大总数。
         */
        /*
        优化：
        剪枝一：保证每个人分一袋饼干
        剪枝二：第一袋饼干，无论分给谁都一样，固定非给第一个人
        剪枝三：如果第 j 个人，j > 1 分得的饼干数量和前一个人一样，跳过
        剪枝四：如果一个人分得的饼干数量 > ans 了，跳过
         */
        Arrays.sort(cookies);
        for (int l = 0, r = cookies.length - 1; l < r; l++, r--) {
            int temp = cookies[l];
            cookies[l] = cookies[r];
            cookies[r] = temp;
        }
        dfs(0, cookies, new int[k], 0);
        return ans;
    }

    int ans = Integer.MAX_VALUE;

    public void dfs(int i, int[] cookies, int[] path, int max) {
        if (i == cookies.length) {
            // 发完了
            if (max < ans) {
                ans = max;
            }
            return;
        }
        int empty = 0;
        for (int x : path) {
            if (x == 0) {
                empty++;
            }
        }
        // 一
        if (empty > cookies.length - i) return;
        for (int j = 0; j < path.length; j++) {
            // 二
            if (i == 0 && j > 0) {
                return;
            }
            // 三
            if (j > 0 && path[j] == path[j -1]) continue;
            // 四
            if (path[j] + cookies[i] > ans) continue;
            path[j] += cookies[i];
            dfs(i + 1, cookies, path, Math.max(path[j], max));
            path[j] -= cookies[i];
        }
    }

}
